File: php/symfony-mysql/app/src/Controller/ArticleController.php
<?php

namespace App\Controller;

use App\Entity\Article;
use App\Repository\ArticleRepository;
use App\Service\NotificationClient;
use OpenTelemetry\API\Metrics\MeterProviderInterface;
use OpenTelemetry\API\Metrics\CounterInterface;
use Psr\Log\LoggerInterface;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Attribute\Route;

#[Route('/api/articles')]
class ArticleController extends AbstractController
{
    private readonly CounterInterface $articlesCreatedCounter;

    public function __construct(
        private readonly ArticleRepository $articleRepository,
        private readonly NotificationClient $notificationClient,
        private readonly LoggerInterface $logger,
        MeterProviderInterface $meterProvider,
    ) {
        $meter = $meterProvider->getMeter('symfony-articles');
        $this->articlesCreatedCounter = $meter->createCounter(
            'articles.created',
            'articles',
            'Number of articles created',
        );
    }

    #[Route('', name: 'article_list', methods: ['GET'])]
    public function list(Request $request): JsonResponse
    {
        $page = max(1, $request->query->getInt('page', 1));
        $perPage = min(100, max(1, $request->query->getInt('per_page', 20)));

        $this->logger->info('Listing articles', ['page' => $page, 'per_page' => $perPage]);

        $result = $this->articleRepository->findPaginated($page, $perPage);
        $data = array_map(fn(Article $a) => $a->toArray(), $result['articles']);

        return new JsonResponse([
            'data' => $data,
            'meta' => [
                'trace_id' => $this->getTraceId(),
                'page' => $page,
                'per_page' => $perPage,
                'total' => $result['total'],
            ],
        ]);
    }

    #[Route('/{id}', name: 'article_show', methods: ['GET'], requirements: ['id' => '\d+'])]
    public function show(int $id): JsonResponse
    {
        $article = $this->articleRepository->find($id);

        if (!$article) {
            $this->logger->warning('Article not found', ['article_id' => $id]);
            return new JsonResponse([
                'error' => ['code' => 'NOT_FOUND', 'message' => 'Article not found'],
                'meta' => ['trace_id' => $this->getTraceId()],
            ], Response::HTTP_NOT_FOUND);
        }

        $this->logger->info('Article retrieved', ['article_id' => $id]);

        return new JsonResponse([
            'data' => $article->toArray(),
            'meta' => ['trace_id' => $this->getTraceId()],
        ]);
    }

    #[Route('', name: 'article_create', methods: ['POST'])]
    public function create(Request $request): JsonResponse
    {
        $payload = json_decode($request->getContent(), true);

        $errors = $this->validate($payload);
        if ($errors) {
            $this->logger->warning('Validation failed', ['errors' => $errors]);
            return new JsonResponse([
                'error' => ['code' => 'VALIDATION_ERROR', 'message' => implode(', ', $errors)],
                'meta' => ['trace_id' => $this->getTraceId()],
            ], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        $article = new Article();
        $article->setTitle($payload['title']);
        $article->setBody($payload['body']);

        $this->articleRepository->save($article);
        $this->articlesCreatedCounter->add(1);
        $this->logger->info('Article created', ['article_id' => $article->getId()]);

        $this->notificationClient->notifyArticleCreated($article->toArray());
        $this->logger->info('Notification sent', ['article_id' => $article->getId()]);

        return new JsonResponse([
            'data' => $article->toArray(),
            'meta' => ['trace_id' => $this->getTraceId()],
        ], Response::HTTP_CREATED);
    }

    #[Route('/{id}', name: 'article_update', methods: ['PUT'], requirements: ['id' => '\d+'])]
    public function update(int $id, Request $request): JsonResponse
    {
        $article = $this->articleRepository->find($id);

        if (!$article) {
            $this->logger->warning('Article not found for update', ['article_id' => $id]);
            return new JsonResponse([
                'error' => ['code' => 'NOT_FOUND', 'message' => 'Article not found'],
                'meta' => ['trace_id' => $this->getTraceId()],
            ], Response::HTTP_NOT_FOUND);
        }

        $payload = json_decode($request->getContent(), true);

        if (!$payload) {
            $this->logger->warning('Validation failed', ['errors' => ['Request body must be valid JSON']]);
            return new JsonResponse([
                'error' => ['code' => 'VALIDATION_ERROR', 'message' => 'Request body must be valid JSON'],
                'meta' => ['trace_id' => $this->getTraceId()],
            ], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        if (isset($payload['title']) && $payload['title'] !== '') {
            if (strlen($payload['title']) > 255) {
                $this->logger->warning('Validation failed', ['errors' => ['title must be 255 characters or less']]);
                return new JsonResponse([
                    'error' => ['code' => 'VALIDATION_ERROR', 'message' => 'title must be 255 characters or less'],
                    'meta' => ['trace_id' => $this->getTraceId()],
                ], Response::HTTP_UNPROCESSABLE_ENTITY);
            }
            $article->setTitle($payload['title']);
        }
        if (isset($payload['body']) && $payload['body'] !== '') {
            $article->setBody($payload['body']);
        }

        $this->articleRepository->save($article);
        $this->logger->info('Article updated', ['article_id' => $id]);

        return new JsonResponse([
            'data' => $article->toArray(),
            'meta' => ['trace_id' => $this->getTraceId()],
        ]);
    }

    #[Route('/{id}', name: 'article_delete', methods: ['DELETE'], requirements: ['id' => '\d+'])]
    public function delete(int $id): JsonResponse
    {
        $article = $this->articleRepository->find($id);

        if (!$article) {
            $this->logger->warning('Article not found for delete', ['article_id' => $id]);
            return new JsonResponse([
                'error' => ['code' => 'NOT_FOUND', 'message' => 'Article not found'],
                'meta' => ['trace_id' => $this->getTraceId()],
            ], Response::HTTP_NOT_FOUND);
        }

        $this->articleRepository->remove($article);
        $this->logger->info('Article deleted', ['article_id' => $id]);

        return new JsonResponse('', Response::HTTP_NO_CONTENT, [], true);
    }

    private function validate(?array $payload): array
    {
        $errors = [];
        if (!$payload) {
            return ['Request body must be valid JSON'];
        }
        if (empty($payload['title'])) {
            $errors[] = 'title is required';
        } elseif (strlen($payload['title']) > 255) {
            $errors[] = 'title must be 255 characters or less';
        }
        if (empty($payload['body'])) {
            $errors[] = 'body is required';
        }
        return $errors;
    }

    private function getTraceId(): string
    {
        $span = \OpenTelemetry\API\Trace\Span::getCurrent();
        $context = $span->getContext();
        return $context->getTraceId();
    }
}

File: java/micronaut-postgres/app/src/main/java/com/example/controller/ArticleController.java
package com.example.controller;

import com.example.model.Article;
import com.example.repository.ArticleRepository;
import com.example.service.NotificationClient;
import com.example.service.TelemetryService;
import io.micronaut.data.model.Page;
import io.micronaut.data.model.Pageable;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.scheduling.TaskExecutors;
import io.micronaut.scheduling.annotation.ExecuteOn;
import io.micronaut.http.annotation.Delete;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.PathVariable;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.annotation.Put;
import io.micronaut.http.annotation.QueryValue;
import io.micronaut.serde.annotation.Serdeable;
import io.opentelemetry.api.trace.Span;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.time.Instant;
import java.util.Map;
import java.util.Optional;

@Controller("/api/articles")
@ExecuteOn(TaskExecutors.BLOCKING)
public class ArticleController {

    private static final Logger LOG = LoggerFactory.getLogger(ArticleController.class);

    private final ArticleRepository articleRepository;
    private final NotificationClient notificationClient;
    private final TelemetryService telemetryService;

    public ArticleController(ArticleRepository articleRepository,
                             NotificationClient notificationClient,
                             TelemetryService telemetryService) {
        this.articleRepository = articleRepository;
        this.notificationClient = notificationClient;
        this.telemetryService = telemetryService;
    }

    @Get
    public HttpResponse<?> list(@QueryValue(defaultValue = "1") int page,
                                @QueryValue(defaultValue = "10") int per_page) {
        Page<Article> result = articleRepository.findAll(Pageable.from(page - 1, per_page));
        return HttpResponse.ok(Map.of(
                "data", result.getContent(),
                "meta", Map.of(
                        "page", page,
                        "per_page", per_page,
                        "total", result.getTotalSize(),
                        "trace_id", currentTraceId()
                )
        ));
    }

    @Get("/{id}")
    public HttpResponse<?> get(@PathVariable Long id) {
        Optional<Article> article = articleRepository.findById(id);
        if (article.isEmpty()) {
            LOG.warn("Article not found: id={}", id);
            return HttpResponse.status(HttpStatus.NOT_FOUND).body(Map.of(
                    "error", Map.of("code", "NOT_FOUND", "message", "Article not found"),
                    "meta", Map.of("trace_id", currentTraceId())
            ));
        }
        return HttpResponse.ok(Map.of(
                "data", article.get(),
                "meta", Map.of("trace_id", currentTraceId())
        ));
    }

    @Post
    public HttpResponse<?> create(@Body CreateArticleRequest request) {
        if (request.title() == null || request.title().isBlank() ||
            request.body() == null || request.body().isBlank()) {
            LOG.warn("Validation failed: title and body are required");
            return HttpResponse.status(HttpStatus.UNPROCESSABLE_ENTITY).body(Map.of(
                    "error", Map.of("code", "VALIDATION_ERROR", "message", "title and body are required"),
                    "meta", Map.of("trace_id", currentTraceId())
            ));
        }

        try {
            Article article = new Article();
            article.setTitle(request.title());
            article.setBody(request.body());
            article = articleRepository.save(article);

            LOG.info("Article created: id={}, title={}", article.getId(), article.getTitle());
            telemetryService.incrementArticlesCreated();

            try {
                notificationClient.notify(Map.of(
                        "id", article.getId(),
                        "title", article.getTitle(),
                        "event", "article.created"
                ));
            } catch (Exception e) {
                LOG.warn("Failed to notify: {}", e.getMessage());
            }

            return HttpResponse.status(HttpStatus.CREATED).body(Map.of(
                    "data", article,
                    "meta", Map.of("trace_id", currentTraceId())
            ));
        } catch (Exception e) {
            LOG.error("Failed to create article: {}", e.getMessage());
            return HttpResponse.serverError(Map.of(
                    "error", Map.of("code", "INTERNAL_ERROR", "message", "Internal server error"),
                    "meta", Map.of("trace_id", currentTraceId())
            ));
        }
    }

    @Put("/{id}")
    public HttpResponse<?> update(@PathVariable Long id, @Body UpdateArticleRequest request) {
        Optional<Article> existing = articleRepository.findById(id);
        if (existing.isEmpty()) {
            LOG.warn("Article not found for update: id={}", id);
            return HttpResponse.status(HttpStatus.NOT_FOUND).body(Map.of(
                    "error", Map.of("code", "NOT_FOUND", "message", "Article not found"),
                    "meta", Map.of("trace_id", currentTraceId())
            ));
        }
        Article article = existing.get();
        if (request.title() != null) article.setTitle(request.title());
        if (request.body() != null) article.setBody(request.body());
        article.setUpdatedAt(Instant.now());
        article = articleRepository.update(article);

        return HttpResponse.ok(Map.of(
                "data", article,
                "meta", Map.of("trace_id", currentTraceId())
        ));
    }

    @Delete("/{id}")
    public HttpResponse<?> delete(@PathVariable Long id) {
        Optional<Article> existing = articleRepository.findById(id);
        if (existing.isEmpty()) {
            LOG.warn("Article not found for delete: id={}", id);
            return HttpResponse.status(HttpStatus.NOT_FOUND).body(Map.of(
                    "error", Map.of("code", "NOT_FOUND", "message", "Article not found"),
                    "meta", Map.of("trace_id", currentTraceId())
            ));
        }
        articleRepository.deleteById(id);
        return HttpResponse.noContent();
    }

    private String currentTraceId() {
        return Span.current().getSpanContext().getTraceId();
    }

    @Serdeable
    public record CreateArticleRequest(String title, String body) {}

    @Serdeable
    public record UpdateArticleRequest(String title, String body) {}
}
